core/field_variable.js

```diff
@@ -192,30 +207,52 @@ Blockly.FieldVariable.prototype.getVariable = function() {
 };
 
 /**
- * Set the variable ID.
- * @param {string} id New variable ID, which must reference an existing
- *     variable.
+ * Gets the validation function for this field, or null if not set.
+ * Returns null if the variable is not set, because validators should not
+ * run on the initial setValue call.
+ * @return {Function} Validation function, or null.
  */
-Blockly.FieldVariable.prototype.setValue = function(id) {
-  var workspace = this.sourceBlock_.workspace;
-  var variable = Blockly.Variables.getVariable(workspace, id);
+Blockly.FieldVariable.prototype.getValidator = function() {
+  // Validators shouldn't operate on the initial setValue call.
+  // Normally this is achieved by calling setValidator after setValue, but
+  // this is not a possibility with variable fields.
+  if (this.variable_) {
+    return this.validator_;
+  }
+};
 
+/**
+ * Ensure that the id belongs to a valid variable of an allowed type.
+ * @param {string} newId The id of the new variable to set.
+ * @return {?string} The validated id, or null if invalid.
+ * @protected
+ */
+Blockly.FieldVariable.prototype.doClassValidation_ = function(newId) {
+  var variable = Blockly.Variables.getVariable(this.workspace_, newId);
   if (!variable) {
-    throw Error('Variable id doesn\'t point to a real variable!  ID was ' + id);
+    console.warn('Variable id doesn\'t point to a real variable! ' +
+        'ID was ' + newId);
+    return null;
   }
-  // Type checks!
+  // Type Checks.
   var type = variable.type;
   if (!this.typeIsAllowed_(type)) {
-    throw Error('Variable type doesn\'t match this field!  Type was ' + type);
-  }
-  if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
-    var oldValue = this.variable_ ? this.variable_.getId() : null;
-    Blockly.Events.fire(new Blockly.Events.BlockChange(
-        this.sourceBlock_, 'field', this.name, oldValue, id));
+    console.warn('Variable type doesn\'t match this field!  Type was ' + type);
+    return null;
   }
-  this.variable_ = variable;
-  this.value_ = id;
-  this.setText(variable.name);
+  return newId;
+};
+
+/**
+ * Update the value of this variable field, as well as its variable and text.
+ * @param {string} newId The id of the new variable.
```

Comment:
The class validator makes sure it's valid, but then any local validator could make it invalid again. This is a problem with all fields. The setValue template method could be changed so that the class validator runs twice if there is a local validator, but there are performance costs.

Reply:
Add a comment that it *should* be valid, and that if validators return bad IDs it could break.

Reply:
Fixed